File: backend/app/schemas/system_settings.py
from pydantic import Field, EmailStr

from .base import BaseSchema
from datetime import datetime, time
from app.models.enums import BackupFrequency

class SettingsUpdateRequest(BaseSchema):
    smtp_host: str | None = None
    smtp_port: int | None = Field(ge=1, le=65535)
    smtp_user: str | None = None
    smtp_password: str | None = None
    smtp_from: EmailStr | None = None
    smtp_use_tls: bool = True

    backup_email: EmailStr | None = None

    backup_enabled: bool = False
    backup_frequency: BackupFrequency = BackupFrequency.MANUAL

    backup_time: time | None = None
    backup_weekday: int | None = Field(default=None, ge=0, le=6)
    backup_monthday: int | None = Field(default=None, ge=1, le=31)

    backup_retention_daily: int = Field(default=30, ge=1)
    backup_retention_weekly: int = Field(default=12, ge=1)
    backup_retention_monthly: int = Field(default=24, ge=1)
    backup_keep_local: bool = True
    backup_send_email: bool = True

    backup_timezone: str = "America/Montevideo"


class SettingsResponse(BaseSchema):

    smtp_host: str | None
    smtp_port: int | None
    smtp_user: str | None

    smtp_from: EmailStr | None
    smtp_use_tls: bool

    smtp_password_configured: bool

    backup_email: EmailStr | None

    backup_enabled: bool
    backup_frequency: BackupFrequency

    backup_time: time | None
    backup_weekday: int | None
    backup_monthday: int | None

    backup_retention_daily: int
    backup_retention_weekly: int
    backup_retention_monthly: int

    backup_keep_local: bool
    backup_send_email: bool

    backup_timezone: str

    last_automatic_backup_at: datetime | None
    next_automatic_backup_at: datetime | None
    last_backup_result: str | None

class EmailTestResponse(BaseSchema):
    success: bool
    sent_to: str

File: frontend/src/types/systemSettings.ts
// types/systemSettings.ts

export const BackupFrequency = {
  MANUAL: "manual",
  DAILY: "daily",
  WEEKLY: "weekly",
  MONTHLY: "monthly"
} as const

export type BackupFrequency =
  typeof BackupFrequency[
    keyof typeof BackupFrequency
  ]


export interface SystemSettings {
  smtp_host: string
  smtp_port: number
  smtp_user: string
  smtp_password: string
  smtp_from: string
  smtp_use_tls: boolean

  backup_email: string

  backup_enabled: boolean
  backup_frequency: BackupFrequency

  backup_retention_daily: number
  backup_retention_weekly: number
  backup_retention_monthly: number

  backup_time: string
  backup_weekday: number | null
  backup_monthday: number | null

  backup_keep_local: boolean
  backup_send_email: boolean

  backup_timezone: string
}

export interface RawSystemSettings {
  smtp_host: string | null
  smtp_port: number
  smtp_user: string | null
  smtp_from: string | null
  smtp_use_tls: boolean
  smtp_password_configured: boolean

  backup_email: string | null

  backup_enabled: boolean
  backup_frequency: BackupFrequency

  backup_time: string

  backup_weekday: number | null
  backup_monthday: number | null

  backup_retention_daily: number
  backup_retention_weekly: number
  backup_retention_monthly: number

  backup_keep_local: boolean
  backup_send_email: boolean

  backup_timezone: string

  last_automatic_backup_at: string | null
  next_automatic_backup_at: string | null
  last_backup_result: string | null
}